frontend/src/scenes/session-recordings/player/snapshot-processing/process-all-snapshots.ts

```diff
@@ -109,8 +153,56 @@ export function processAllSnapshots(
                 hasSeenMeta = true
             }
 
+            {
+                const windowId = snapshot.windowId
+                const hasSeenFullForWindow = !!seenFullByWindow[windowId]
+
+                if (
+                    snapshot.type === EventType.IncrementalSnapshot &&
+                    !hasSeenFullForWindow &&
+                    isLikelyMobileScreenshot(snapshot)
+                ) {
+                    // Inject a synthetic full snapshot (and meta if needed) immediately before the first incremental
+                    const syntheticTimestamp = Math.max(0, snapshot.timestamp - 1)
+
+                    if (!hasSeenMeta) {
+                        const viewport = viewportForTimestamp(syntheticTimestamp)
+                        if (viewport && viewport.width && viewport.height) {
+                            const metaEvent: RecordingSnapshot = {
+                                type: EventType.Meta,
+                                timestamp: syntheticTimestamp,
+                                windowId: snapshot.windowId,
+                                data: {
+                                    width: parseInt(viewport.width, 10),
+                                    height: parseInt(viewport.height, 10),
+                                    href: viewport.href || 'unknown',
+                                },
+                            }
+                            result.push(metaEvent)
+                            sourceResult.push(metaEvent)
+                            throttleCapture(`${sessionRecordingId}-patched-meta`, () => {
+                                posthog.capture('patched meta into web recording', {
+                                    throttleCaptureKey: `${sessionRecordingId}-patched-meta`,
+                                    sessionRecordingId,
+                                    sourceKey: sourceKey,
+                                    feature: 'session-recording-meta-patching',
+                                })
+                            })
+                        }
+                    }
+
+                    const syntheticFull = createMinimalFullSnapshot(snapshot.windowId, syntheticTimestamp)
```

Comment:
if you literally convert the meta you're looking at into a full snapshot
then the check below will pick it up and patch the meta in without needing the duplication

(there may be reasons not to do that, in which case let's pull the duplication out as a method - it's already very long to read in here 🙈)

otherwise i think looking good